### PR TITLE
Add new command "kubectl set selector"

### LINKED ISSUE #7296: kubectl: enable nodeSelector updates from the command line
We saw in the flocker demo this command:

master$ kubectl get rc redis-master -o yaml | sed 's/spinning/ssd/' | kubectl update -f -

Should be able to do this with a kubectl command, not spitting out yaml, sed, and reading it back in.

### PR DESCRIPTION
**Pull request by asalkeld**
This is also to aid PR #27109 (create subcommands)

Part of the new set commands: #21648

``` release-note
Added new command kubectl set selector
```

Related: #7296

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28949)

### CODE REVIEW

**Comment by asalkeld** [comment]
@kubernetes/kubectl

File: pkg/kubectl/cmd/set/set_selector.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+		Long:    fmt.Sprintf(selectorLong),
+		Example: selectorExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.Complete(f, cmd, args))
+			cmdutil.CheckErr(options.Validate())
+			cmdutil.CheckErr(options.Run())
+		},
+	}
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().Bool("all", false, "select all resources in the namespace of the specified resource types")
```
Comment (kargakis) [inline pkg/kubectl/cmd/set/set_selector.go]: Capitalize the first letter

File: pkg/kubectl/cmd/util/factory.go [hunk 2]
```diff
@@ -767,6 +770,27 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
...
+			case *extensions.DaemonSet:
+				return fn(nil, t.Spec.Selector)
+			case *extensions.ReplicaSet:
+				return fn(nil, t.Spec.Selector)
+			case *batch.Job:
+				return fn(nil, t.Spec.Selector)
+			case *api.PersistentVolumeClaim:
+				return fmt.Errorf("the object does not allow updates to the selector")
+			default:
+				return fmt.Errorf("the object is does not have a selector")
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/factory.go]: remove "is"

File: pkg/kubectl/cmd/set/set_selector.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+		// patch the change
+		obj, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, api.StrategicMergePatchType, patch.Patch)
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("failed to patch selector update to template: %v\n", err))
+			continue
+		}
+		info.Refresh(obj, true)
+
+		// record this change (for rollout history)
```
Comment (kargakis) [inline pkg/kubectl/cmd/set/set_selector.go]: Why not add this change at the same time we are doing the patch?

Comment (asalkeld) [inline pkg/kubectl/cmd/set/set_selector.go]: @kargakis to be honest I copied this little chunk from set_image.go, so no smart reason :-O

Comment (kargakis) [inline pkg/kubectl/cmd/set/set_selector.go]: @janetkuo got any idea?

File: pkg/kubectl/cmd/set/set_selector.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, o.Recursive, o.Filenames...).
+		Flatten()
+	if !o.Local {
+		builder = builder.
+			ResourceTypeOrNameArgs(o.All, o.Resources...).
+			Latest()
+	}
+	o.Infos, err = builder.Do().Infos()
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: This might make remote calls.  Try to refactor so the remote lookup happens in `Run`

File: pkg/kubectl/cmd/set/set_selector.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+	Filenames   []string
+	Recursive   bool
+	ShortOutput bool
+	Overwrite   bool
+	All         bool
+	Record      bool
+	ChangeCause string
+	Local       bool
+	DryRun      bool
+	Cmd         *cobra.Command
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: Part of the reason for this `Options` structure is for future composition that probably won't have all the things you expect on the `cmd`.  I think you can factor `PrintObject` to be a function that doesn't take a command and construct that function inside of `Complete`.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+	"k8s.io/kubernetes/pkg/runtime"
+	utilerrors "k8s.io/kubernetes/pkg/util/errors"
+)
+
+// SelectorOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type SelectorOptions struct {
+	Mapper      meta.RESTMapper
+	Typer       runtime.ObjectTyper
+	Infos       []*resource.Info
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: This should be local to `Run`

File: pkg/kubectl/cmd/set/set_selector.go [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+			}
+		}
+
+		info.Refresh(obj, true)
+		cmdutil.PrintSuccess(o.Mapper, o.ShortOutput, o.Out, info.Mapping.Resource, info.Name, "selector updated")
+	}
+	return utilerrors.NewAggregate(allErrs)
+}
+
+// getResourcesAndSelector retrieves resources and the selector expression from the given args (assuming selectors the last arg)
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: I'm not super familiar with the new label selector, but I can see that the lexer has some logic around whitespace.  Do we have another example of entering it in the CLI?  I'm not entirely sure that "single arg at the end" is going to be easiest option given that we optionally specify resources here.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+					return err
+				}
+			}
+
+			return nil
+		})
+		return true, err
+	})
+
+	for _, patch := range patches {
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: I think I'd like to see this method use `.Visit` like most of the other CLI commands that deal with resources from a builder.

File: pkg/kubectl/cmd/util/factory.go [hunk 9]
```diff
@@ -145,6 +145,9 @@ type Factory struct {
 	// UpdatePodSpecForObject will call the provided function on the pod spec this object supports,
 	// return false if no pod spec is supported, or return an error.
 	UpdatePodSpecForObject func(obj runtime.Object, fn func(*api.PodSpec) error) (bool, error)
+	// UpdateSelectorForObject calls fn with that applicable selector type to update.
+	UpdateSelectorForObject func(obj runtime.Object, fn func(*map[string]string, *unversioned.LabelSelector) error) error
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: I should be able to co-erce every map based selector into an `unversioned.LabelSelector` and simply use that, right?  Types that only support one can validate that the set is legal.

File: pkg/kubectl/cmd/util/factory.go [hunk 10]
```diff
@@ -767,6 +770,27 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 				return false, fmt.Errorf("the object is not a pod or does not have a pod template")
 			}
 		},
+		// UpdateSelectorForObject update the selector for the provided object
+		UpdateSelectorForObject: func(obj runtime.Object, fn func(*map[string]string, *unversioned.LabelSelector) error) error {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: I'm ok with adding this, but its time to start building reflective tests around it.  Write a test that installs all the types, navigates the scheme to check them all for an `unversioned.LabelSelector` field, then makes sure that doesn't end up with unsupported error from the switch.

Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: Seems like your usage actually just needs to get back the ref to the selector and then you can do whatever you want to it.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("failed to patch selector update to template: %v\n", err))
+			continue
+		}
+		info.Refresh(obj, true)
+
+		// record this change (for rollout history)
+		if o.Record || cmdutil.ContainsChangeCause(info) {
+			if err := cmdutil.RecordChangeCause(obj, o.ChangeCause); err == nil {
+				if obj, err = resource.NewHelper(info.Client, info.Mapping).Replace(info.Namespace, info.Name, false, obj); err != nil {
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: You'll want to do this using a patch like `edit` does.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+					allErrs = append(allErrs, fmt.Errorf("changes to %s/%s can't be recorded: %v\n", info.Mapping.Resource, info.Name, err))
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: This shouldn't be an error, you fulfilled your mission, you just couldn't update the hint.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+			cmdutil.CheckErr(options.Complete(f, cmd, args))
+			cmdutil.CheckErr(options.Validate())
+			cmdutil.CheckErr(options.Run())
+		},
+	}
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().Bool("all", false, "Select all resources in the namespace of the specified resource types")
+	cmd.Flags().String("resource-version", "", "If non-empty, the selectors update will only succeed if this is the current resource-version for the object. Only valid when specifying a single resource.")
+	usage := "Filename, directory, or URL to a file identifying the resource to update the selectors"
+	cmd.Flags().Bool("overwrite", false, "If true, allow selectors to be overwritten, otherwise reject selector updates that overwrite existing selectors.")
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: I think we're almost always going to be overwriting, I don't see a need for this flag.

**Comment by deads2k** [comment]
Great idea for a command.

File: pkg/kubectl/cmd/set/set_selector.go (hunk 7 above)
Comment (asalkeld) [inline pkg/kubectl/cmd/set/set_selector.go]: @deads2k most commands support "-l selector" which is the same syntax.
```
func (b *Builder) SelectorParam(s string) *Builder {
	selector, err := labels.Parse(s)
...
func ParseToLabelSelector(selector string) (*LabelSelector, error) {
	reqs, err := labels.ParseToRequirements(selector)
 ParseToRequirements() just calls the same underlying parse function. So I was hoping that this would be the least surprise to users. Although maybe it could be under the "-l selector" but I thought that might be even more confusing (like this was to first select on objects before setting them). Open to ideas...

File: pkg/kubectl/cmd/util/factory.go (hunk 9 above)
Comment (asalkeld) [inline pkg/kubectl/cmd/util/factory.go]: The problem is that the old api objects use map[string]string and the newer ones use the new struct, so I need a way of returning the supported type from the api objects.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+	}
+	return nil
+}
+
+// Run the command.
+func (o *SelectorOptions) Run() error {
+	allErrs := []error{}
+
+	patches := CalculatePatches(o.Infos, o.Encoder, func(info *resource.Info) (bool, error) {
+		err := o.UpdateSelectorForObject(info.Object, func(map_selector *map[string]string, selector *unversioned.LabelSelector) error {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/set/set_selector.go]: style: use camelCase for all variables, no underscores except in extremely rare cases (generated methods in conversions and deep copies are our only exceptions today)

File: pkg/kubectl/cmd/set/set_selector.go [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+			if map_selector != nil {
+				if len(o.Selector.MatchExpressions) > 0 {
+					return fmt.Errorf("error: MatchExpression %v not supported on this object", o.Selector.MatchExpressions)
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/set/set_selector.go]: Should not have to prefix errors with "error:", which should be handled by the generic cmd framework, and in general don't use `CamelCase` in messages (if this is a flag, prefer `--match-expression`, or just use `match expression`.

File: pkg/kubectl/cmd/util/factory.go (hunk 9 above)
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/factory.go]: It would be better for your fn to return the new selector or an error, rather than passing by reference.  Returning the new value for "value types" is preferred generally.

Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: > The problem is that the old api objects use map[string]string and the newer ones use the new struct, so I need a way of returning the supported type from the api objects.

Pretty sure that `LabelSelector.MatchLabels` is the equivalent.  You should be able to detect whether you have any `.MatchExpression`.  If you don't, then its assignable.  If you do, then you fail.  Same logic either way, its just a question of whether we keep a datatype we don't want anymore in this code, right?

**Comment by asalkeld** [comment]
Thanks all for the reviews, I am working on a new patch that will hopefully be better.

File: pkg/kubectl/cmd/set/helper.go [hunk 16]
```diff
@@ -112,40 +112,49 @@ type Patch struct {
 	Patch  []byte
 }
 
+// CalculatePatch calls the mutation function on the provided info object, and generates a strategic merge patch for
+// the changes in the object. Encoder must be able to encode the info into the appropriate destination type.
+// This function returns whatever mutateFn returns.
+func CalculatePatch(patch *Patch, encoder runtime.Encoder, mutateFn func(*resource.Info) (bool, error)) bool {
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/helper.go]: Can this take an `Info` and return `[]byte, error` rather than mutating input?  that gives us a more "normal" API?

File: pkg/kubectl/cmd/set/set_selector.go [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+		return copyNewSelector(t.Spec.Selector)
+	case *api.PersistentVolumeClaim:
+		return fmt.Errorf("the object does not allow updates to the selector")
+	default:
+		return fmt.Errorf("the object does not have a selector")
+	}
+}
+
+// getResourcesAndSelector retrieves resources and the selector expression from the given args (assuming selectors the last arg)
+func getResourcesAndSelector(args []string) (resources []string, selector *unversioned.LabelSelector, err error) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: Please avoid named return values.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+		if len(selector.MatchExpressions) > 0 {
+			return nil, fmt.Errorf("match expression %v not supported on this object", selector.MatchExpressions)
+		}
+		dst = make(map[string]string)
+		for label, value := range selector.MatchLabels {
+			dst[label] = value
+		}
+		return dst, nil
+	}
+	copyNewSelector := func(dst *unversioned.LabelSelector) error {
```
Comment (deads2k) [inline pkg/kubectl/cmd/set/set_selector.go]: return a new `unversioned.LabelSelector` instead of overwriting the input arg.

**Comment by deads2k** [comment]
@kubernetes/rh-ux @fabianofranz this is going to set a pattern for updating selectors on general objects in the CLI.

**Comment by bgrant0607** [comment]
@asalkeld @deads2k 

Note that this doesn't actually fix #7296, which was about setting nodeSelector, so I changed the PR description to just reference that issue rather than close it. There was a confusing comment in the issue, which I just deleted.

More importantly, this is very likely to orphan pods and/or create overlapping selectors, which probably is not what the user intended. I'm not keen to make it easier for users to shoot themselves in the foot/torso/head.

It's important to think about the use case. I believe a common use case is:

1. User creates a controller and service.
2. Later, user wants to perform rolling update or create a canary, but didn't add a label to the original whose value could be changed in the new controller.

One solution could be a command to add a label to a controller's pod template and to all of the pods before also adding the label to the label selector. That's pretty tricky to do correctly. It should be a library routine that could be used in all the places we need to do that, such as `kubectl rolling-update --image`. We'd want to add fullyLabeledReplicas to the controllers that don't yet support it.

Another solution, for controllers supporting the new selector, is to update the selector of the original controller to exclude pods with the new label key (using DoesNotExist). That wouldn't orphan its existing pods, and isn't as race-prone as adding labels to existing pods. 

This command could be used to do that, but the chance of users figuring out that trick on their own before doing something damaging is low.

What did you intend to do with this?

**Comment by deads2k** [comment]
> What did you intend to do with this?

I was thinking primarily about changing the selector of a `service` for a "copy this service, change selector, steal an RC's pod, monkey around with that particular pod" debugging flow.  I hadn't considered the RC pod orphaning problem, but RC's that ensure that the selector selects the labels on the `PodTemplateSpec`.  You're worried about cases where someone went to the trouble of changing pod labels and didn't stamp out new pods?

**Comment by asalkeld** [comment]
@bgrant0607 also @deads2k had the idea of using this command in a pipeline like this: https://github.com/kubernetes/kubernetes/issues/25382#issuecomment-222672195
One option is to only allow LocalOnly changes, so it assists the user to generate the template to send to "kubectl create -f".

**Comment by bgrant0607** [comment]
@deads2k Yes, we do validate that controller selectors match their pod templates, which probably means that `set selector` would usually fail. :-) At that point, a likely scenario is that the user would change both the pod template and selector, without relabeling existing pods, which would orphan them.

Regarding the debugging use case: The usual pattern is to remove or change a label, which would cause it to be replaced by its controller and removed from its service, disable liveness probes (which we don't have an easy way to do AFAIK), use the proxy or port forwarding to contact it, and use logs and exec to debug it. Why is a new service needed? How would the pod be removed from the original service and controller?

@asalkeld Thanks, that helps explain the use case.

If we were to support controllers as well as services, we'd need really good help and documentation explaining the pitfalls.

**Comment by deads2k** [comment]
Go ahead and split out `--local` on `kubectl label`.  That ought to be non-contentious and great to have.

I also think that @bgrant0607 gave an "ok" for this, but would like to see it restricted to services to start, then perhaps expanded to other selecting types with more help.

**Comment by bgrant0607** [comment]
@deads2k characterized my position accurately.

**Comment by asalkeld** [comment]
So I have split out the --local on label. I'll look at the restricting it to services.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+var (
+	selectorLong = dedent.Dedent(`
+		Update the selectors on a resource.
+
+		A selector must begin with a letter or number, and may contain letters, numbers, hyphens, dots, and underscores, up to %[1]d characters.
+		If --resource-version is specified, then updates will use this resource version, otherwise the existing resource-version will be used.`)
+	selectorExample = dedent.Dedent(`
+                # set the labels and selector before creating a deployment/service pair.
+                kubectl create service cluster-ip -o yaml --dry-run | kubectl set selector --local -f - 'environment=qa' -o yaml | kubectl create -f - 
+                kubectl create deployment my-dep -o yaml --dry-run | kubectl label --local -f - environment=qa -o yaml | kubectl create -f -
```
Comment (kargakis) [inline pkg/kubectl/cmd/set/set_selector.go]: Make sure the initial examples will be using services only.

File: pkg/kubectl/cmd/set/set_selector.go [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+	Resources       []string
+	Selector        *unversioned.LabelSelector
+	PrintObject     func(mapper meta.RESTMapper, obj runtime.Object, out io.Writer) error
+}
+
+var (
+	selectorLong = dedent.Dedent(`
+		Update the selectors on a resource.
+
+		A selector must begin with a letter or number, and may contain letters, numbers, hyphens, dots, and underscores, up to %[1]d characters.
```
Comment (kargakis) [inline pkg/kubectl/cmd/set/set_selector.go]: Add a stanza about only services being supported for now,

File: pkg/kubectl/cmd/set/set_selector.go [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+			dst.MatchLabels[label] = value
+		}
+
+		dst.MatchExpressions = make([]unversioned.LabelSelectorRequirement, len(selector.MatchExpressions))
+		copy(dst.MatchExpressions, selector.MatchExpressions)
+		return &dst
+	}
+	var err error
+	switch t := obj.(type) {
+	case *api.ReplicationController:
```
Comment (kargakis) [inline pkg/kubectl/cmd/set/set_selector.go]: Should be supporting only service for now.

**Comment by asalkeld** [comment]
Thanks @kargakis , new patch coming soon.

**Comment by kargakis** [comment]
@asalkeld can you also add a test in hack/make-rules/test-cmd.sh?

**Comment by asalkeld** [comment]
@kargakis I have added the test into test-cmd.sh and raised the issue.

**Comment by asalkeld** [comment]
@k8s-bot gci gke e2e test this

File: pkg/kubectl/cmd/set/set_selector.go [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+	if err != nil {
+		return err
+	}
+
+	o.changeCause = f.Command()
+	mapper, typer := f.Object()
+	o.mapper = mapper
+	o.encoder = f.JSONEncoder()
+
+	o.builder = resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), f.Decoder(true)).
```
Comment (kargakis) [inline pkg/kubectl/cmd/set/set_selector.go]: You can now use the factory to create the resource builder for you. Try f.NewBuilder()

File: pkg/kubectl/cmd/set/helper.go (hunk 16 above)
Comment (kargakis) [inline pkg/kubectl/cmd/set/helper.go]: Agreed. @asalkeld can you make this change?

**Comment by kargakis** [comment]
Two last comments and then I am tagging this.

File: pkg/kubectl/cmd/set/set_selector.go (hunk 22 above)
Comment (asalkeld) [inline pkg/kubectl/cmd/set/set_selector.go]: OK, will do.

File: pkg/kubectl/cmd/set/helper.go (hunk 16 above)
Comment (asalkeld) [inline pkg/kubectl/cmd/set/helper.go]: OK, I'll give it a go. It will make the PR bigger..

File: pkg/kubectl/cmd/rollout/rollout_pause.go [hunk 23]
```diff
@@ -38,7 +38,7 @@ import (
 type PauseConfig struct {
 	resource.FilenameOptions
 
-	Pauser  func(info *resource.Info) (bool, error)
+	Pauser  func(info *resource.Info) ([]byte, error)
```
Comment (kargakis) [inline pkg/kubectl/cmd/rollout/rollout_pause.go]: Why this change? Seems unrelated to this PR

File: pkg/kubectl/cmd/rollout/rollout_resume.go [hunk 24]
```diff
@@ -38,7 +38,7 @@ import (
 type ResumeConfig struct {
 	resource.FilenameOptions
 
-	Resumer func(object *resource.Info) (bool, error)
+	Resumer func(object *resource.Info) ([]byte, error)
```
Comment (kargakis) [inline pkg/kubectl/cmd/rollout/rollout_resume.go]: Why this change? Seems unrelated to this PR

Comment (asalkeld) [inline pkg/kubectl/cmd/rollout/rollout_resume.go]: Hi @kargakis thanks for coming back to this PR. This was asked for here: https://github.com/kubernetes/kubernetes/pull/28949#r72980078
this needs to change as it is passed into the CalculatePatch() function.

Comment (kargakis) [inline pkg/kubectl/cmd/rollout/rollout_resume.go]: I see, thanks for the clarification

**Comment by asalkeld** [comment]
@k8s-bot node e2e test this

**Comment by asalkeld** [comment]
Hi folks, trying to get this in (my last day in Mirantis is tomorrow - not sure how to handle the CLA after that point if it doesn't get in).

**Comment by asalkeld** [comment]
@bgrant0607 @deads2k ptal

**Comment by bgrant0607** [comment]
@kargakis User-facing changes need release notes. I'm changing the release-note label.

**Comment by bgrant0607** [comment]
@asalkeld Unfortunately we're past the feature freeze for 1.5, so I don't think this can be merged right now.

**Comment by asalkeld** [comment]
ok thanks.

**Comment by kargakis** [comment]
@asalkeld I can take this over and make sure it gets merged if that's ok with you.

**Comment by asalkeld** [comment]
Hi Michail, please do. I am not working on k8s anymore (change of work
situation).

Angus

**Comment by kargakis** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/38966
